File: src/pytypes/onefuzztypes/events.py
#!/usr/bin/env python
#
# Copyright (c) Microsoft Corporation.
# Licensed under the MIT License.

from datetime import datetime
from enum import Enum
from typing import Any, Dict, List, Optional, Union
from uuid import UUID, uuid4

from pydantic import BaseModel, Field

from ._monkeypatch import _check_hotfix
from .enums import (
    OS,
    Architecture,
    NodeState,
    ScalesetState,
    TaskState,
    TaskType,
    VmState,
)
from .models import (
    AutoScaleConfig,
    Error,
    InstanceConfig,
    JobConfig,
    RegressionReport,
    Report,
    TaskConfig,
    UserInfo,
)
from .primitives import Container, PoolName, Region
from .responses import BaseResponse


class BaseEvent(BaseModel):
    pass


class EventTaskStopped(BaseEvent):
    job_id: UUID
    task_id: UUID
    user_info: Optional[UserInfo]
    config: TaskConfig


class EventTaskFailed(BaseEvent):
    job_id: UUID
    task_id: UUID
    error: Error
    user_info: Optional[UserInfo]
    config: TaskConfig


class EventJobCreated(BaseEvent):
    job_id: UUID
    config: JobConfig
    user_info: Optional[UserInfo]


class JobTaskStopped(BaseModel):
    task_id: UUID
    task_type: TaskType
    error: Optional[Error]


class EventJobStopped(BaseEvent):
    job_id: UUID
    config: JobConfig
    user_info: Optional[UserInfo]
    task_info: Optional[List[JobTaskStopped]]


class EventTaskCreated(BaseEvent):
    job_id: UUID
    task_id: UUID
    config: TaskConfig
    user_info: Optional[UserInfo]


class EventTaskStateUpdated(BaseEvent):
    job_id: UUID
    task_id: UUID
    state: TaskState
    end_time: Optional[datetime]
    config: TaskConfig


class EventTaskHeartbeat(BaseEvent):
    job_id: UUID
    task_id: UUID
    config: TaskConfig


class EventPing(BaseEvent, BaseResponse):
    ping_id: UUID


class EventScalesetCreated(BaseEvent):
    scaleset_id: UUID
    pool_name: PoolName
    vm_sku: str
    image: str
    region: Region
    size: int


class EventScalesetFailed(BaseEvent):
    scaleset_id: UUID
    pool_name: PoolName
    error: Error


class EventScalesetDeleted(BaseEvent):
    scaleset_id: UUID
    pool_name: PoolName


class EventScalesetResizeScheduled(BaseEvent):
    scaleset_id: UUID
    pool_name: PoolName
    size: int


class EventPoolDeleted(BaseEvent):
    pool_name: PoolName


class EventPoolCreated(BaseEvent):
    pool_name: PoolName
    os: OS
    arch: Architecture
    managed: bool
    autoscale: Optional[AutoScaleConfig]


class EventProxyCreated(BaseEvent):
    region: Region
    proxy_id: Optional[UUID]


class EventProxyDeleted(BaseEvent):
    region: Region
    proxy_id: Optional[UUID]


class EventProxyFailed(BaseEvent):
    region: Region
    proxy_id: Optional[UUID]
    error: Error


class EventProxyStateUpdated(BaseEvent):
    region: Region
    proxy_id: UUID
    state: VmState


class EventNodeCreated(BaseEvent):
    machine_id: UUID
    scaleset_id: Optional[UUID]
    pool_name: PoolName


class EventNodeHeartbeat(BaseEvent):
    machine_id: UUID
    scaleset_id: Optional[UUID]
    pool_name: PoolName


class EventNodeDeleted(BaseEvent):
    machine_id: UUID
    scaleset_id: Optional[UUID]
    pool_name: PoolName


class EventScalesetStateUpdated(BaseEvent):
    scaleset_id: UUID
    pool_name: PoolName
    state: ScalesetState


class EventNodeStateUpdated(BaseEvent):
    machine_id: UUID
    scaleset_id: Optional[UUID]
    pool_name: PoolName
    state: NodeState


class EventCrashReported(BaseEvent):
    report: Report
    container: Container
    filename: str
    task_config: Optional[TaskConfig]


class EventRegressionReported(BaseEvent):
    regression_report: RegressionReport
    container: Container
    filename: str
    task_config: Optional[TaskConfig]


class EventFileAdded(BaseEvent):
    container: Container
    filename: str


class EventInstanceConfigUpdated(BaseEvent):
    config: InstanceConfig


Event = Union[
    EventJobCreated,
    EventJobStopped,
    EventNodeStateUpdated,
    EventNodeCreated,
    EventNodeDeleted,
    EventNodeHeartbeat,
    EventPing,
    EventPoolCreated,
    EventPoolDeleted,
    EventProxyFailed,
    EventProxyCreated,
    EventProxyDeleted,
    EventProxyStateUpdated,
    EventScalesetFailed,
    EventScalesetCreated,
    EventScalesetDeleted,
    EventScalesetStateUpdated,
    EventScalesetResizeScheduled,
    EventTaskFailed,
    EventTaskStateUpdated,
    EventTaskCreated,
    EventTaskStopped,
    EventTaskHeartbeat,
    EventCrashReported,
    EventRegressionReported,
    EventFileAdded,
    EventInstanceConfigUpdated,
]


class EventType(Enum):
    job_created = "job_created"
    job_stopped = "job_stopped"
    node_created = "node_created"
    node_deleted = "node_deleted"
    node_state_updated = "node_state_updated"
    ping = "ping"
    pool_created = "pool_created"
    pool_deleted = "pool_deleted"
    proxy_created = "proxy_created"
    proxy_deleted = "proxy_deleted"
    proxy_failed = "proxy_failed"
    proxy_state_updated = "proxy_state_updated"
    scaleset_created = "scaleset_created"
    scaleset_deleted = "scaleset_deleted"
    scaleset_failed = "scaleset_failed"
    scaleset_state_updated = "scaleset_state_updated"
    scaleset_resize_scheduled = "scaleset_resize_scheduled"
    task_created = "task_created"
    task_failed = "task_failed"
    task_state_updated = "task_state_updated"
    task_stopped = "task_stopped"
    crash_reported = "crash_reported"
    regression_reported = "regression_reported"
    file_added = "file_added"
    task_heartbeat = "task_heartbeat"
    node_heartbeat = "node_heartbeat"
    instance_config_updated = "instance_config_updated"


EventTypeMap = {
    EventType.job_created: EventJobCreated,
    EventType.job_stopped: EventJobStopped,
    EventType.node_created: EventNodeCreated,
    EventType.node_deleted: EventNodeDeleted,
    EventType.node_state_updated: EventNodeStateUpdated,
    EventType.node_heartbeat: EventNodeHeartbeat,
    EventType.ping: EventPing,
    EventType.pool_created: EventPoolCreated,
    EventType.pool_deleted: EventPoolDeleted,
    EventType.proxy_created: EventProxyCreated,
    EventType.proxy_deleted: EventProxyDeleted,
    EventType.proxy_failed: EventProxyFailed,
    EventType.proxy_state_updated: EventProxyStateUpdated,
    EventType.scaleset_created: EventScalesetCreated,
    EventType.scaleset_deleted: EventScalesetDeleted,
    EventType.scaleset_failed: EventScalesetFailed,
    EventType.scaleset_state_updated: EventScalesetStateUpdated,
    EventType.scaleset_resize_scheduled: EventScalesetResizeScheduled,
    EventType.task_created: EventTaskCreated,
    EventType.task_failed: EventTaskFailed,
    EventType.task_state_updated: EventTaskStateUpdated,
    EventType.task_heartbeat: EventTaskHeartbeat,
    EventType.task_stopped: EventTaskStopped,
    EventType.crash_reported: EventCrashReported,
    EventType.regression_reported: EventRegressionReported,
    EventType.file_added: EventFileAdded,
    EventType.instance_config_updated: EventInstanceConfigUpdated,
}


def get_event_type(event: Event) -> EventType:

    for (event_type, event_class) in EventTypeMap.items():
        if isinstance(event, event_class):
            return event_type

    raise NotImplementedError("unsupported event type: %s" % type(event))


class EventMessage(BaseEvent):
    event_id: UUID = Field(default_factory=uuid4)
    event_type: EventType
    event: Event
    instance_id: UUID
    instance_name: str


# because Pydantic does not yet have discriminated union types yet, parse events
# by hand.  https://github.com/samuelcolvin/pydantic/issues/619
def parse_event_message(data: Dict[str, Any]) -> EventMessage:
    instance_id = UUID(data["instance_id"])
    instance_name = data["instance_name"]
    event_id = UUID(data["event_id"])
    event_type = EventType[data["event_type"]]
    # mypy incorrectly identifies this as having not supported parse_obj yet
    event = EventTypeMap[event_type].parse_obj(data["event"])  # type: ignore

    return EventMessage(
        event_id=event_id,
        event_type=event_type,
        event=event,
        instance_id=instance_id,
        instance_name=instance_name,
    )


_check_hotfix()

File: src/ApiService/ApiService/OneFuzzTypes/Events.cs
using System.Reflection;
using System.Text.Json;
using System.Text.Json.Serialization;
using Microsoft.OneFuzz.Service.OneFuzzLib.Orm;

namespace Microsoft.OneFuzz.Service;

/// <summary>
/// Identifies the enum type associated with the event class
/// </summary>
[AttributeUsage(AttributeTargets.Class, AllowMultiple = false, Inherited = false)]
public class EventTypeAttribute : Attribute {
    public EventTypeAttribute(EventType eventType) {
        this.EventType = eventType;
    }

    public EventType EventType { get; }
}


public enum EventType {
    JobCreated,
    JobStopped,
    NodeCreated,
    NodeDeleted,
    NodeStateUpdated,
    Ping,
    PoolCreated,
    PoolDeleted,
    ProxyCreated,
    ProxyDeleted,
    ProxyFailed,
    ProxyStateUpdated,
    ScalesetCreated,
    ScalesetDeleted,
    ScalesetFailed,
    ScalesetStateUpdated,
    ScalesetResizeScheduled,
    TaskCreated,
    TaskFailed,
    TaskStateUpdated,
    TaskStopped,
    CrashReported,
    RegressionReported,
    FileAdded,
    TaskHeartbeat,
    NodeHeartbeat,
    InstanceConfigUpdated,
    NotificationFailed,
    WebhookSucceeded,
    WebhookRetried,
    WebhookFailed
}

public abstract record BaseEvent() {
    private static readonly IReadOnlyDictionary<Type, EventType> _typeToEvent;
    private static readonly IReadOnlyDictionary<EventType, Type> _eventToType;

    static BaseEvent() {
        static EventType ExtractEventType(Type type) {
            var attr = type.GetCustomAttribute<EventTypeAttribute>();
            if (attr is null) {
                throw new InvalidOperationException($"Type {type} is missing {nameof(EventTypeAttribute)}");
            }
            return attr.EventType;
        }

        _typeToEvent =
            typeof(BaseEvent).Assembly.GetTypes()
            .Where(t => t.IsSubclassOf(typeof(BaseEvent)))
            .ToDictionary(x => x, ExtractEventType);

        _eventToType = _typeToEvent.ToDictionary(x => x.Value, x => x.Key);

        // check that all event types are accounted for
        var allEventTypes = Enum.GetValues<EventType>();
        var missingEventTypes = allEventTypes.Except(_eventToType.Keys).ToList();
        if (missingEventTypes.Any()) {
            throw new InvalidOperationException($"Missing event types: {string.Join(", ", missingEventTypes)}");
        }
    }


    public EventType GetEventType() {
        var type = this.GetType();
        if (_typeToEvent.TryGetValue(type, out var eventType)) {
            return eventType;
        }

        throw new NotSupportedException($"Unknown event type: {type.GetType()}");
    }

    public static Type GetTypeInfo(EventType eventType) {
        if (_eventToType.TryGetValue(eventType, out var type)) {
            return type;
        }

        throw new ArgumentException($"Unknown event type: {eventType}");
    }
};

public class EventTypeProvider : ITypeProvider {
    public Type GetTypeInfo(object input) {
        return BaseEvent.GetTypeInfo((input as EventType?) ?? throw new ArgumentException($"input is expected to be an EventType {input}"));
    }
}

[EventType(EventType.TaskStopped)]
public record EventTaskStopped(
    Guid JobId,
    Guid TaskId,
    StoredUserInfo? UserInfo,
    TaskConfig Config
) : BaseEvent();

[EventType(EventType.TaskFailed)]
public record EventTaskFailed(
    Guid JobId,
    Guid TaskId,
    Error Error,
    StoredUserInfo? UserInfo,
    TaskConfig Config
    ) : BaseEvent();


[EventType(EventType.JobCreated)]
public record EventJobCreated(
   Guid JobId,
   JobConfig Config,
   StoredUserInfo? UserInfo
   ) : BaseEvent();


public record JobTaskStopped(
    Guid TaskId,
    TaskType TaskType,
    Error? Error
    );


[EventType(EventType.JobStopped)]
public record EventJobStopped(
    Guid JobId,
    JobConfig Config,
    StoredUserInfo? UserInfo,
    List<JobTaskStopped> TaskInfo
) : BaseEvent(), ITruncatable<BaseEvent> {
    public BaseEvent Truncate(int maxLength) {
        return this with {
            Config = Config.Truncate(maxLength)
        };
    }
}


[EventType(EventType.TaskCreated)]
public record EventTaskCreated(
    Guid JobId,
    Guid TaskId,
    TaskConfig Config,
    StoredUserInfo? UserInfo
    ) : BaseEvent();

[EventType(EventType.TaskStateUpdated)]
public record EventTaskStateUpdated(
    Guid JobId,
    Guid TaskId,
    TaskState State,
    DateTimeOffset? EndTime,
    TaskConfig Config
    ) : BaseEvent();

[EventType(EventType.TaskHeartbeat)]
public record EventTaskHeartbeat(
   Guid JobId,
   Guid TaskId,
   string? Project,
   string? Name,
   TaskState? State,
   TaskConfig? Config
) : BaseEvent();

[EventType(EventType.Ping)]
public record EventPing(
    Guid PingId
) : BaseEvent();


[EventType(EventType.ScalesetCreated)]
public record EventScalesetCreated(
   ScalesetId ScalesetId,
   PoolName PoolName,
   string VmSku,
   string Image,
   Region Region,
   int Size) : BaseEvent();


[EventType(EventType.ScalesetFailed)]
public sealed record EventScalesetFailed(
    ScalesetId ScalesetId,
    PoolName PoolName,
    Error Error
) : BaseEvent();


[EventType(EventType.ScalesetDeleted)]
public record EventScalesetDeleted(
   ScalesetId ScalesetId,
   PoolName PoolName

   ) : BaseEvent();


[EventType(EventType.ScalesetResizeScheduled)]
public record EventScalesetResizeScheduled(
    ScalesetId ScalesetId,
    PoolName PoolName,
    long size
    ) : BaseEvent();


[EventType(EventType.PoolDeleted)]
public record EventPoolDeleted(
   PoolName PoolName
   ) : BaseEvent();


[EventType(EventType.PoolCreated)]
public record EventPoolCreated(
   PoolName PoolName,
   Os Os,
   Architecture Arch,
   bool Managed
   // ignoring AutoScaleConfig because it's not used anymore
   //AutoScaleConfig? Autoscale
   ) : BaseEvent();


[EventType(EventType.ProxyCreated)]
public record EventProxyCreated(
   Region Region,
   Guid? ProxyId
   ) : BaseEvent();


[EventType(EventType.ProxyDeleted)]
public record EventProxyDeleted(
   Region Region,
   Guid? ProxyId
) : BaseEvent();


[EventType(EventType.ProxyFailed)]
public record EventProxyFailed(
   Region Region,
   Guid? ProxyId,
   Error Error
) : BaseEvent();


[EventType(EventType.ProxyStateUpdated)]
public record EventProxyStateUpdated(
   Region Region,
   Guid ProxyId,
   VmState State
   ) : BaseEvent();


[EventType(EventType.NodeCreated)]
public record EventNodeCreated(
    Guid MachineId,
    ScalesetId? ScalesetId,
    PoolName PoolName
    ) : BaseEvent();

[EventType(EventType.NodeHeartbeat)]
public record EventNodeHeartbeat(
    Guid MachineId,
    ScalesetId? ScalesetId,
    PoolName PoolName,
    NodeState state
    ) : BaseEvent();


[EventType(EventType.NodeDeleted)]
public record EventNodeDeleted(
    Guid MachineId,
    ScalesetId? ScalesetId,
    PoolName PoolName,
    NodeState? MachineState
) : BaseEvent();


[EventType(EventType.ScalesetStateUpdated)]
public record EventScalesetStateUpdated(
    ScalesetId ScalesetId,
    PoolName PoolName,
    ScalesetState State
) : BaseEvent();

[EventType(EventType.NodeStateUpdated)]
public record EventNodeStateUpdated(
    Guid MachineId,
    ScalesetId? ScalesetId,
    PoolName PoolName,
    NodeState state
    ) : BaseEvent();

[EventType(EventType.CrashReported)]
public record EventCrashReported(
    Report Report,
    Container Container,
    [property: JsonPropertyName("filename")] String FileName,
    TaskConfig? TaskConfig
) : BaseEvent(), ITruncatable<BaseEvent> {
    public BaseEvent Truncate(int maxLength) {
        return this with {
            Report = Report.Truncate(maxLength)
        };
    }
}

[EventType(EventType.RegressionReported)]
public record EventRegressionReported(
    RegressionReport RegressionReport,
    Container Container,
    [property: JsonPropertyName("filename")] String FileName,
    TaskConfig? TaskConfig
) : BaseEvent(), ITruncatable<BaseEvent> {
    public BaseEvent Truncate(int maxLength) {
        return this with {
            RegressionReport = RegressionReport.Truncate(maxLength)
        };
    }
}

[EventType(EventType.FileAdded)]
public record EventFileAdded(
    Container Container,
    [property: JsonPropertyName("filename")] String FileName
) : BaseEvent();


[EventType(EventType.InstanceConfigUpdated)]
public record EventInstanceConfigUpdated(
    InstanceConfig Config
) : BaseEvent();

[EventType(EventType.NotificationFailed)]
public record EventNotificationFailed(
    Guid NotificationId,
    Guid JobId,
    Error? Error
) : BaseEvent();

[EventType(EventType.WebhookSucceeded)]
public record EventWebhookSucceeded(
    Guid WebhookId,
    Guid EventId,
    EventType eventType,
    string? data
) : BaseEvent();

[EventType(EventType.WebhookRetried)]
public record EventWebhookRetried(
    Guid WebhookId,
    Guid EventId,
    EventType eventType,
    string? data
) : BaseEvent();

[EventType(EventType.WebhookFailed)]
public record EventWebhookFailed(
    Guid WebhookId,
    Guid EventId,
    EventType eventType,
    string? data
) : BaseEvent();

public record DownloadableEventMessage : EventMessage, ITruncatable<DownloadableEventMessage> {
    public Uri SasUrl { get; init; }
    public DateOnly? ExpiresOn { get; init; }

    public DownloadableEventMessage(Guid EventId, EventType EventType, BaseEvent Event, Guid InstanceId, string InstanceName, DateTime CreatedAt, Uri SasUrl, DateOnly? ExpiresOn)
        : base(EventId, EventType, Event, InstanceId, InstanceName, CreatedAt) {
        this.SasUrl = SasUrl;
        this.ExpiresOn = ExpiresOn;
    }

    public override DownloadableEventMessage Truncate(int maxLength) {
        if (this.Event is ITruncatable<BaseEvent> truncatableEvent) {
            return this with {
                Event = truncatableEvent.Truncate(maxLength)
            };
        } else {
            return this;
        }
    }
}

public record EventMessage(
    Guid EventId,
    EventType EventType,
    [property: TypeDiscrimnator("EventType", typeof(EventTypeProvider))]
    [property: JsonConverter(typeof(BaseEventConverter))]
    BaseEvent Event,
    Guid InstanceId,
    String InstanceName,
    DateTime CreatedAt,
    String Version = "1.0"
) : ITruncatable<EventMessage>, IRetentionPolicy {
    public virtual EventMessage Truncate(int maxLength) {
        if (this.Event is ITruncatable<BaseEvent> truncatableEvent) {
            return this with {
                Event = truncatableEvent.Truncate(maxLength)
            };
        } else {
            return this;
        }
    }

    public DateOnly GetExpiryDate() => Event switch {
        BaseEvent @event when @event is IRetentionPolicy retentionPolicy => retentionPolicy.GetExpiryDate(),
        _ => DateOnly.FromDateTime(DateTime.UtcNow.AddDays(180))
    };
}

public class BaseEventConverter : JsonConverter<BaseEvent> {
    public override BaseEvent? Read(ref Utf8JsonReader reader, Type typeToConvert, JsonSerializerOptions options) {
        throw new NotSupportedException("BaseEvent cannot be read");
    }

    public override void Write(Utf8JsonWriter writer, BaseEvent value, JsonSerializerOptions options) {
        var eventType = value.GetType();
        JsonSerializer.Serialize(writer, value, eventType, options);
    }
}
